### ISSUE TITLE
Feature request: if commit messages have [USJ-XXX], auto-link

### ISSUE THREAD

**Issue by rplevy** [FundingCircle/shipment_tracker#211]
Feature request: if commit messages have [USJ-XXX]-style Jira identifiers, auto-link the PR to that Jira in shipment tracker.

**Comment by DanBurton** [comment]
Or if the PR title has it.

**Comment by sgerrand** [comment]
@rplevy @DanBurton: Can you think of a more generic feature request? e.g.

> If a commit message contains a `[JIRAPROJECT-XXX]` reference string, attempt to automatically link the pull request which contains that commit to the JIRA ticket.

**Comment by sgerrand** [comment]
> if commit messages have [USJ-XXX]-style Jira identifiers, auto-link the PR to that Jira in shipment tracker.

I can see some potential issues with this change, namely that you could potentially have multiple pull requests with the same issue tracker reference identifier in their commit sets. If the relationship between an issue tracker reference and a pull request must be one to one, how should this be resolved?

**Comment by DanBurton** [comment]
>  If the relationship between an issue tracker reference and a pull request must be one to one

Is this our current assumption?

A single issue sometimes corresponds to multiple PRs.
A single PR sometimes (but rarely) corresponds to multiple issues.

If the auto-link feature has problems with this, then I think a "first come first served" resolution strategy is simple and effective. It can auto-link the first time it sees a relevant commit, and not link any subsequent relevant commits. Those who open a PR should still be checking that shipment tracker has the relevant links. This proposed feature is a mere convenience that also encourages commits to include this useful info.

**Comment by damien** [comment]
I agree with @DanBurton—while many PRs are one-to-one, it's not unusual to have a unit of work that is related to multiple PR (or vice-versa). There are also situations where we have things like dups, bugs with the same root cause, etc that force a one-to-many relationship.

**Comment by sgerrand** [comment]
>> If the relationship between an issue tracker reference and a pull request must be one to one
>
> Is this our current assumption?

That's [currently correct](https://github.com/FundingCircle/shipment_tracker/blob/fd2ac41a567c14730e0221a3931eb6eb836f3922/features/releases.feature).

**Comment by sgerrand** [comment]
> while many PRs are one-to-one, it's not unusual to have a unit of work that is related to multiple PR (or vice-versa).

Sub tasks are a good way to maintain the relationship between the elements which comprise a unit of work. (Side note: _should_ a unit of work comprise multiple pull requests? My initial 💭  was that it shouldn't.)

> There are also situations where we have things like dups, bugs with the same root cause, etc that force a one-to-many relationship.

Again, that sounds like a problem which can be solved in an issue tracker rather than Shipment Tracker. Issue trackers generally always support linking issues and in the cases presented, they _should_ be linked there.

**Comment by sgerrand** [comment]
> If the auto-link feature has problems with this, then I think a "first come first served" resolution strategy is simple and effective. It can auto-link the first time it sees a relevant commit, and not link any subsequent relevant commits. Those who open a PR should still be checking that shipment tracker has the relevant links. This proposed feature is a mere convenience that also encourages commits to include this useful info.

This seems reasonable to me.

@FundingCircle/technical-leads: What do you think?

**Comment by damien** [comment]
^ @sgerrand That's more or less what my team does.

> Sub tasks are a good way to maintain the relationship between the elements which comprise a unit of work. (Side note: should a unit of work comprise multiple pull requests? My initial 💭 was that it shouldn't.)

I think this is heavily dependent on the workflow of a particular team. My antidotal experience has shown me that it's much easier to manage linked tickets than it is to manage tickets with subtasks.

> Again, that sounds like a problem which can be solved in an issue tracker rather than Shipment Tracker. Issue trackers generally always support linking issues and in the cases presented, they should be linked there.

I initially disagreed with this, as I was thinking there were cases where ST could block deployment of  work that was covered in multiple PRs or multiple tickets, but after thinking it through I think you're correct:
* In cases where multiple PRs map to a single ticket, everything works fine—we just need to do a bit of manual linking between ST and JIRA.
* For cases where multiple tickets reference the same PR, manual linking works fine as well.
* If the most common use-case is 1-1 JIRA issue/PR, the logic as described here lines up nicely.

**Comment by creese** [comment]
> Sub tasks are a good way to maintain the relationship between the elements which comprise a unit of work.

JIRA sub-tasks can't be pointed. As such, we don't use them.

> the relationship between an issue tracker reference and a pull request must be one to one

I don't think this is compatible with the requirement that master should reflect production. If we aren't ready to deploy, we usually merge to a feature branch. If the branch contains a new application, this can go on for weeks or even months.

**Comment by kylejm** [comment]
May be worth checking #412, which has now been merged and deployed
